### PR TITLE
Engine traits cleanup

### PR DESCRIPTION
**Pull request by brson**
Polish up the engine_traits and engine_rocks crates to use consistent organization, naming, and patterns. Add docs.

###  What have you changed?

This is mostly taking what I've learned and implemented in https://github.com/tikv/tikv/pull/5657 and doing the same thing consistently across the engine_traits and engine_rocks traits, and documenting what I'm doing so others can do the same.

This extends https://github.com/tikv/tikv/pull/5657, so most of the commits are from that PR. New commits start at "move db.rs to engine.rs".

###  What is the type of the changes?

- Engineering (engineering change which doesn't change any feature or fix any issue)

###  How is the PR tested?

typechecking

###  Does this PR affect `tidb-ansible`?

no

###  Refer to a related PR or issue link (optional)

https://github.com/tikv/tikv/issues/4184

### CODE REVIEW

**Comment by brson** [comment]
I'll rebase this tomorrow.

**Comment by brson** [comment]
OK, this is rebased against master and now only includes new commits.

**Comment by brson** [comment]
ptal @5kbpers @aknuds1 @overvenus

File: components/engine_rocks/src/lib.rs [hunk 1]
```diff
@@ -1,170 +1,44 @@
 // Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
 
+//! Implementation of engine_traits for RocksDB
+//!
+//! This is a work-in-progress attempt to abstract all the features needed by
+//! TiKV to persist its data.
+//!
+//! The module structure here mirrrors that in engine_traits where possible.
```
Comment (kennytm) [inline components/engine_rocks/src/lib.rs]:
```suggestion
//! The module structure here mirrors that in engine_traits where possible.
```

File: components/engine_traits/src/lib.rs [hunk 2]
```diff
@@ -1,79 +1,108 @@
...
-mod options;
-pub use crate::options::*;
-mod db_options;
-mod util;
-pub use crate::db_options::*;
+// These modules contain traits that need to be implemented by engines, either
+// they are required by KvEngine or are an associated type of KvEngine. It is
+// recommended that engines follow the same module layout.
+//
+// Many of these define "extension" traits, that end in `Ex`.
```
Comment (kennytm) [inline components/engine_traits/src/lib.rs]:
```suggestion
// Many of these define "extension" traits, that end in `Ext`.
```

File: components/engine_rocks/src/engine.rs [hunk 3]
```diff
@@ -108,47 +80,31 @@ impl KvEngine for Rocks {
     fn cf_names(&self) -> Vec<&str> {
         self.0.cf_names()
     }
-
-    fn delete_all_in_range_cf(
```
Comment (Connor1996) [inline components/engine_rocks/src/engine.rs]: why removing this?

File: components/engine_rocks/src/tests.rs [hunk 4]
```diff
@@ -0,0 +1,144 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
```
Comment (Connor1996) [inline components/engine_rocks/src/tests.rs]: Is it necessary to move the tests to a separate file? seems it's not a TiKV convention.

File: components/engine_traits/src/import.rs [hunk 5]
```diff
@@ -34,3 +28,9 @@ pub trait Import: CFHandleExt {
         expected_checksum: u32,
     ) -> Result<()>;
 }
+
+pub trait IngestExternalFileOptions {
```
Comment (5kbpers) [inline components/engine_traits/src/import.rs]: Shall we convert other options to traits to keep consistency in this PR?

**Comment by zhangjinpeng1987** [comment]
Please address comments.

File: components/engine_rocks/src/engine.rs (hunk 3 above)
Comment (brson) [reply components/engine_rocks/src/engine.rs]: From the commit message:

> The code is unused, is duplicated between engine_rocks and engine and is not the
    same. I'm going to come back to this when it actually needs to be used.

I'm suspicious of it.

File: components/engine_traits/src/import.rs (hunk 5 above)
Comment (zhangjinpeng1987) [reply components/engine_traits/src/import.rs]: Maybe step by step is better.

File: components/engine_traits/src/lib.rs [hunk 6]
```diff
@@ -1,79 +1,108 @@
...
-    }
+pub const DATA_KEY_PREFIX_LEN: usize = 1;
 
-    pub fn sync_raft(&self) -> Result<()> {
-        self.raft.sync()
-    }
-}
+// In our tests, we found that if the batch size is too large, running delete_all_in_range will
+// reduce OLTP QPS by 30% ~ 60%. We found that 32K is a proper choice.
+pub const MAX_DELETE_BATCH_SIZE: usize = 32 * 1024;
```
Comment (zhangjinpeng1987) [inline components/engine_traits/src/lib.rs]: I think this is a RocksDB relative variable, better to move to engine_rocks.

File: components/engine_rocks/src/tests.rs (hunk 4 above)
Comment (brson) [reply components/engine_rocks/src/tests.rs]: I've moved them to engine.rs.

Comment (zhangjinpeng1987) [reply components/engine_rocks/src/tests.rs]: I think it is ok.

File: components/engine_traits/src/lib.rs (hunk 6 above)
Comment (brson) [reply components/engine_traits/src/lib.rs]: The const is duplicated from the engine crate and appears to be unused so I've just deleted it.

**Comment by brson** [comment]
Comments addressed.

File: components/engine_rocks/src/engine.rs (hunk 3 above)
Comment (zhangjinpeng1987) [reply components/engine_rocks/src/engine.rs]: https://github.com/tikv/tikv/blob/master/src/raftstore/store/fsm/apply.rs#L1293 and https://github.com/tikv/tikv/blob/master/src/server/gc_worker.rs#L391 use delete_all_in_range_cf.

**Comment by Connor1996** [comment]
`delete_all_in_range_cf` is still removed from engine trait?

**Comment by brson** [comment]
@Connor1996 I explained this a few places: I removed it because it was duplicating a function from the engine crate, while having a slightly different implementation, and no callers. It was untested code, so I removed it in favor of the version that was tested. When callers are converted to use the trait, then the implementation can be moved from the engine crate to the engine_rocks crate, leaving no code duplication, and ensuring the function stays correct.

**Comment by Connor1996** [comment]
@brson Oh sorry, didn't notice that. I was thought we should copy the implementation from the `engine` crate to the `engine_rocks` crate, so we can keep the `delete_all_in_range_cf` in the trait. And once the callers are converted to use the trait, we remove the `engine` crate totally.
What I want to express is we need it in the trait, so remove it now and then add it back is unnecessary. Anyway, it's fine for me after your explanation. 😉Good job!

**Pull request merged**